### ISSUE TITLE
[INFRA/CORE] Logging actions in previous turn

### ISSUE THREAD

**Issue by Darrekt** [SOMAS2020/SOMAS2020#56]
In this issue we propose a modification to `GameState` that allows us to store the actions taken in each turn.

As a quick refresher to those who may not have read the code, presently the game runs in a loop, triggered by the server's `EntryPoint` method. The observed `state` is simply an array of `GameState` objects.

```go
// EntryPoint function that returns a list of historic common.GameStates until the
// game ends.
func (s *SOMASServer) EntryPoint() ([]common.GameState, error) {
	states := []common.GameState{s.gameState.Copy()}

	for !s.gameOver(config.MaxTurns, config.MaxSeasons) {
		if err := s.runTurn(); err != nil {
			return states, err
		}
		states = append(states, s.gameState.Copy())
	}
	return states, nil
}
```

The proposition is to add a `map[turnNumber][]common.Action` to the GameState at each turn. This can be implemented in DispatchActions, where every action that is successfully dispatched and reduced is collected in a `slice` and then added onto the modified `GameState` at the end.

A PR will be following sometime today.

**Comment by lhl2617** [comment]
I'm thinking it'd be nicer to have `map[turnNumber]common.ActionResult` where 
```
type ActionResult {
    Action common.Action // the action
    Result error // the result of the action
}
```
so we can keep track of errored out actions instead of ignoring them. This also means that clients can check if their actions succeeded.

**Comment by lhl2617** [comment]
Can you also ask IIGO/IIFO/IITO team if their actions have dedicated `initiators` (i.e. which island initiated it?) then we can extend it to `map[turnNumber](map[clientID]common.ActionResult)` which makes it even easier to track -- essentially we group them by which `clientID` initiated it. This is straightforward for end of turn actions, but idk if IIGO/IIFO/IITO has dedicated initiators for actions, or e.g. in IIGO it's a `collective` action. 

Even so we could make an enum spanning all possible initiators

**Comment by lhl2617** [comment]
cc @Neelesh99  @777jameskim

**Comment by lhl2617** [comment]
I'm overhauling the actions interface in favour of a better one in #59 - please work off that branch/merge that in first.

**Comment by Darrekt** [comment]
I've merged it and fixed something that it broke in #61.

Am I having a brain fart, or did we just not think it through and blindly say that we needed a `map[TurnNumber]gamestate.ActionResult` when it should instead just be a `[]gamestate.ActionResult`? Each `GameState` is already mapped to (has) a specific `Turn`. 

Am I missing something? If not, I'll go ahead and implement this as discussed with `ActionResult` with a slice.

**Comment by lhl2617** [comment]
closing for now, actions is now obsolete
